fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/expressions/OneOfThemWithComparison.java

```diff
@@ -102,23 +101,26 @@ public QueryComponent withOtherComparison(Comparisons.Comparison comparison) {
     }
 
     @Override
-    public ExpandedPredicates normalizeForPlanner(@Nonnull final CorrelationIdentifier baseAlias, @Nonnull final List<String> fieldNamePrefix) {
+    public GraphExpansion expand(@Nonnull final CorrelationIdentifier baseAlias, @Nonnull final List<String> fieldNamePrefix) {
         List<String> fieldNames = ImmutableList.<String>builder()
                 .addAll(fieldNamePrefix)
                 .add(getFieldName())
                 .build();
         final Quantifier childBase = Quantifier.forEach(GroupExpressionRef.of(new ExplodeExpression(baseAlias, fieldNames)));
-        final SelectExpression selectExpression = ExpandedPredicates.ofPredicate(new ValuePredicate(
-                new ObjectValue(childBase.getAlias()), comparison)).buildSelectWithBase(childBase);
+        final SelectExpression selectExpression =
+                GraphExpansion.ofPredicate(new QuantifiedObjectValue(childBase.getAlias()).withComparison(comparison)).buildSelectWithBase(childBase);
         final Quantifier.Existential childQuantifier = Quantifier.existential(GroupExpressionRef.of(selectExpression));
 
+        // create a query component that creates a path to this prefix and then applies this to it
```

Comment:
In general compensation is computed by comparing two graphs. In fact this will lead us to the need of reapplying what the outcome of a particular subgraph is. Since we don't have any runtime for anything `QueryComponent`-like outside of the `QueryComponent` tree I need to keep this alternative version around -- yuck!

Here is what happens in this case specifically: We could just reapply the original query filter. I order to be a bit better I can also (if I have to) reapply what's needed that is originally under this expansion, e.g.

query: `Field(a).oneOfThem(Field(b).oneOfThem(c = 5) and d = 6)`

If I have an index for `a, d` but nothing for `a,b,c` I can still use the index on `a, d` and reapply either the entire query filter or this one:

 `Field(a).oneOfThem(Field(b).oneOfThem(c = 5))`

The code in question takes care of making a chain of `Field(a).oneOfThem(Field(b))....`. Note that the last step is just the current key expression at hand.

Reply:
✅